File: solutions/code-carl/ch09_dp/T13_LC0518_Change.java
/**
 * Practice template for:
 * - LeetCode 518. Coin Change II
 *   https://leetcode.cn/problems/coin-change-ii/
 */
/**
 * @author Cheng Yihao
 * @version 1.0
 * @date 2024/11/22 19:51
 * @comment
 */
public class T13_LC0518_Change {
    public int change(int amount, int[] coins) {
        int[] dp = new int[amount + 1];
        dp[0] = 1;
        for (int coin : coins) {
            for (int j = coin; j <= amount; j++) {
                dp[j] += dp[j - coin];
            }
        }
        return dp[amount];
    }
}

File: templates/python/code-carl/ch09_dp/t13_lc0518_change.py
"""LeetCode 518. Coin Change II
https://leetcode.cn/problems/coin-change-ii/
"""

from typing import List


class Solution:
    def change(self, amount: int, coins: List[int]) -> int:
        pass

if __name__ == "__main__":
    solution = Solution()
    assert solution.change(5, [1, 2, 5]) == 4
    assert solution.change(3, [2]) == 0
    assert solution.change(10, [10]) == 1
    print("All tests passed.")
